r/R/dplyr-functions.R

```diff
@@ -672,6 +672,16 @@ nse_funcs$strptime <- function(x, format = "%Y-%m-%d %H:%M:%S", tz = NULL, unit
   Expression$create("strptime", x, options = list(format = format, unit = unit))
 }
 
+nse_funcs$strftime <- function(x, format = "", tz = "", usetz = FALSE) {
+  if (tz != "") {
+    arrow_not_supported("tz argument")
```

Comment:
Can we cast the arrow array here? Ideally we would simply cast it to the desired timezone (metadata change).

Reply:
What casting do you refer to?

Reply:
Changing the type from e.g.: `timestamp("us", "CET")` to `timestamp("us", tz)` where `tz` is the argument of the `strftime` function.

Reply:
I am not in favor of casting here and would let the C++ implementation handle the given inputs. This would ensure a more consistent behavior across R/Python/C++.

Reply:
The way `strftime` kernel is currently implemented in arrow is that the timezone is taken from metadata of the timestamp array. (There is a [jira](https://issues.apache.org/jira/browse/ARROW-13247) for a timezone changing kernel.)

I think this is consistent with most `strftime` implementations (e.g. [pandas](https://pandas.pydata.org/docs/reference/api/pandas.Series.dt.strftime.html)) but obviously not this one.

I'd prefer to either do casting of the timestamp here (I'm not sure it's technically possible. @nic - do you know if it is?) or error out.
Another option is of course to alter the c++ `strftime`, I'd then make a jira and we can discuss it there.

Reply:
Is it possible? Maybe.  Here's the steps I'd take to find out:

1. Open R, load Arrow, and call `list_compute_functions()` to get a list of compute kernels available from R.  I see "cast" is there, so this looks promising.  Alternatively, I could start by looking at the C++ compute kernels documentation to see if there are any fitting kernels.

2. Open `arrow/r/src/compute.cpp` to check if the `cast` kernel has its options bound and if so, what are they.  It does, and the key one here is `to_type`.

3. Play around with `call_function()` (R function you can use for directly calling compute kernels to test things etc).  I ended up with this example working for me: `call_function("cast", Array$create(c(1:3L)), options = list(to_type = utf8()))`.  Not the simil

4. I'm pretty sure we can have Expressions which take other Expressions as arguments, I've implemented things with this in the past, though there are no current code examples, so I'd experiment with that.  Here's an example using `call_function()` which could be analogous to how it works with Expressions:

```
call_function(
  "replace_substring",
  call_function(
    "cast",
    Array$create(c(1:3L)),
    options = list(to_type = utf8())
  ),
  options = list(pattern = "1", replacement = "one")
)
```

Should we do it?  I *think* so, though I'm a bit hesitant to say "yes" as I'm concerned about unintended consequences or inconsistencies.  But then again, I guess that's what unit tests are for, right? ;)

Reply:
Thanks @thisisnic, this is of great help! :)

Reply:
@thisisnic casting approach works, thanks again for the pointers.

New question - can we get type from from within dplyr functions? We'd need so as to not convert timestamps between units when we cast. What I'd ideally do is `x$type$unit()` but it fails.

Reply:
`unit <- TimestampType__unit(x$type())` did the trick.